wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevFileConverter.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.String.format;
+import static org.eclipse.che.api.devfile.server.Constants.CURRENT_SPEC_VERSION;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+import org.eclipse.che.api.devfile.model.Action;
+import org.eclipse.che.api.devfile.model.Command;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.model.Project;
+import org.eclipse.che.api.devfile.model.Source;
+import org.eclipse.che.api.devfile.model.Tool;
+import org.eclipse.che.api.workspace.server.model.impl.CommandImpl;
+import org.eclipse.che.api.workspace.server.model.impl.ProjectConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.SourceStorageImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+
+/** Helps to convert Devfile into workspace config and back. */
+public class DevFileConverter {
+
+  public Devfile workspaceToDevFile(WorkspaceConfigImpl wsConfig) {
```

Comment:
What about converting workspace config which contains recipe? As far as I understand it should be unsupported for time being, but this method designed not to throw an exception and doesn't have the corresponding check.

Reply:
Which exception do you propose to throw? `UnsupportedOperation` ? And in which case? If there is `defaultEnv` set and/or map of environments is not empty ?

Reply:
@mshaposhnik 
> Which exception do you propose to throw? `UnsupportedOperation` ?

`UnsupportedOperation` definitely not. I would propose to introduce some new checked exception. 

> `And in which case? If there is defaultEnv set and/or map of environments is not empty ?`

If there is any information in workspace config that can not be stored in Devfile. Like any `environment` is configured now,  `multiply-environments` later.
In other words if `WorkspaceConfig 1` -> `Devfile` -> `WorkspaceConfig 2` and there is no ability to provide `WorkspaceConfig 1` == `WorkspaceConfig 2`
I may be wrong here @skabashnyuk WDYT?